src/go/configgenerator/filtergen/grpc_field_extraction.go

```diff
@@ -0,0 +1,171 @@
+	confpb "google.golang.org/genproto/googleapis/api/serviceconfig"
+	smpb "google.golang.org/genproto/googleapis/api/servicemanagement/v1"
+	"google.golang.org/protobuf/proto"
+	protov2 "google.golang.org/protobuf/proto"
+	descpb "google.golang.org/protobuf/types/descriptorpb"
+
+	policypb "google3/google/api/policy_go_proto"
+	corepb "google3/third_party/envoy/src/api/envoy/config/core/v3/pkg_go_proto"
+)
+
+const (
+	// GRPCFieldExtractionFilterName is the debug name of the extension.
+	GRPCFieldExtractionFilterName = "envoy.filters.http.grpc_field_extraction"
+)
+
+// GRPCFieldExtractionGenerator is a filter generator for the filter config.
+type GRPCFieldExtractionGenerator struct {
+	// ProtoDescriptorBin is the serialized proto descriptor set.
+	ProtoDescriptorBin []byte
+
+	// PoliciesBySelector maps from OP selector to OSS MethodPolicy.
+	PoliciesBySelector map[string]*policypb.MethodPolicy
+
+	NoopFilterGenerator
+}
+
+// NewGRPCFieldExtractionFilterGensFromOPConfig creates a GRPCFieldExtractionGenerator from
+// OP service config + descriptor + ESPv2 options. It is a FilterGeneratorOPFactory.
+func NewGRPCFieldExtractionFilterGensFromOPConfig(serviceConfig *confpb.Service, opts options.ConfigGeneratorOptions) ([]FilterGenerator, error) {
+	grpcGen, err := NewGRPCFieldExtractionFilterGenFromOPConfig(serviceConfig, opts, true)
+	if err != nil {
+		return nil, err
+	}
+	if grpcGen == nil {
+		return nil, nil
+	}
+
+	return []FilterGenerator{
+		grpcGen,
+	}, nil
+}
+
+// NewGRPCFieldExtractionFilterGenFromOPConfig creates a single GRPCFielcExtractorGenerator.
+//
+// It also has the option to skip the filter. If enabled, all checks will occur
+// and filter may not be created. Otherwise, filter is always created.
+func NewGRPCFieldExtractionFilterGenFromOPConfig(serviceConfig *confpb.Service, opts options.ConfigGeneratorOptions, maybeSkipFilter bool) (*GRPCFieldExtractionGenerator, error) {
+	var descriptorBin []byte
+	foundDescriptor := false
+
+	for _, sourceFile := range serviceConfig.GetSourceInfo().GetSourceFiles() {
+		configFile := &smpb.ConfigFile{}
+		err := sourceFile.UnmarshalTo(configFile)
+		if err != nil {
+			continue
+		}
+
+		if configFile.GetFileType() == smpb.ConfigFile_FILE_DESCRIPTOR_SET_PROTO {
+			foundDescriptor = true
+			descriptorBin = configFile.GetFileContents()
+		}
+	}
+
+	// Cannot check `descriptorBin == nil` because many tests use empty descriptor.
+	if !foundDescriptor {
+		// b/148605552: Previous versions of the `gcloud_build_image` script did not download the proto descriptor.
+		// We cannot ensure that users have the latest version of the script, so notify them via non-fatal logs.
+		// Log as error instead of warning because error logs will show up even if `--enable_debug` is false.
+		glog.Error("Unable to setup gRPC Field Extraction because no proto descriptor was found in the service config. " +
+			"Please use version 2020-01-29 (or later) of the `gcloud_build_image` script. " +
```

Comment:
We can probably break this behavior now, it's been 3 years :)

Let's change it to an error like "Unable to setup gRPC field extraction because no proto descriptor was found in the service config."